source/common/filter/tcp_proxy.cc

```diff
@@ -122,11 +116,30 @@ TcpProxyStats TcpProxyConfig::generateStats(const std::string& name, Stats::Scop
                               POOL_GAUGE_PREFIX(scope, final_prefix))};
 }
 
+void TcpProxy::finalizeUpstreamConnectionStats() {
+  read_callbacks_->upstreamHost()->cluster().stats().upstream_cx_destroy_.inc();
+  read_callbacks_->upstreamHost()->cluster().stats().upstream_cx_active_.dec();
+  read_callbacks_->upstreamHost()->stats().cx_active_.dec();
+  read_callbacks_->upstreamHost()
+      ->cluster()
+      .resourceManager(Upstream::ResourcePriority::Default)
+      .connections()
+      .dec();
+  connected_timespan_->complete();
+}
+
+void TcpProxy::closeUpstreamConnection() {
+  finalizeUpstreamConnectionStats();
+  upstream_connection_->close(Network::ConnectionCloseType::NoFlush);
+  read_callbacks_->connection().dispatcher().deferredDelete(std::move(upstream_connection_));
+}
+
 void TcpProxy::initializeReadFilterCallbacks(Network::ReadFilterCallbacks& callbacks) {
   read_callbacks_ = &callbacks;
   ENVOY_CONN_LOG(debug, "new tcp proxy session", read_callbacks_->connection());
 
   read_callbacks_->connection().addConnectionCallbacks(downstream_callbacks_);
+  read_callbacks_->connection().readDisable(true);
```

Comment:
Can we add a small comment on the necessity of readDisable() here? (Presumably so we don't write to an upstream that might fail).